rfcs/027-first-class-queries.md

```diff
@@ -0,0 +1,81 @@
+- Feature Name: first_class_queries
+- Start Date: 2021-03-18
+- RFC PR: [amundsen-io/rfcs#0000](https://github.com/amundsen-io/rfcs/pull/27)
+- Amundsen Issue: [amundsen-io/amundsen#0000](https://github.com/amundsen-io/amundsen/issues/27)
+
+# First Class Queries
+
+## Summary
+
+This proposes to ingest user-run queries directly into Amundsen. The information included would be the raw string query (typically SQL), as well as the objects referenced by it (Columns, Tables, etc). Amundsen would not parse SQL natively, rather it would provide the plumbing for users who have SQL parsing capabilities.
+
+## Motivation
+
+This allows for considerably more dynamism in how Amundsen surfaces usage-based information. Rather than popular tables being determined primarily by querying usage statistics from dashboarding systems, Amundsen could determine it at runtime by querying the metastore. Same goes for frequent table users. It also lays the groundwork for features like: commonly joined tables, popular queries, audit logging, etc. Additionally, if DDL/DML changes are observed in the log, they can be associated with a table.
+
+## Guide-level Explanation (aka Product Details)
+
+In Amundsen, query logs from your DWH, RDBMS or BI tool may be ingested raw. It's up to you to extract the information about which subjects a particular query touches (e.g. which tables it queries, which user ran it, etc), but once ingested, Amundsen can surface interesting insights about query patterns.
+
+## UI/UX-level Explanation
+
+This change does not necessarily entail UX changes. Most simply, we could surface all queries run against a table (similarly to how dashboard queries are currently surfaced). Mostly, though, this underpins future features, as mentioned in Motivation.
+
+## Reference-level Explanation (aka Technical Details)
+
+### Existing functionality
+
+This feature is intended to be implemented entirely separately from existing storage of usage patterns, at least initially. Once we get the implementation in place, we may choose to start using it as the normalized source of truth for more usage information.
+
+This would interact with `DashboardQuery`, in that this would be a superset of that functionality. I suggest implementing this first, and transitioning existing `DashboardQuery` code to use this more generic functionality.
+
+Eventually, this could subsume the denormalized form `TableColumnUsage` for installs that have full data available. It is not likely we would want to deprecate `TableColumnUsage` to allow for different types of integrations. It's possible we would want to pre-generate `TableColumnUsage` from the underlying data, both for compatibility and read-time performance reasons.
+
+### databuilder
+
+We would create a new `class Query(GraphSerializable)` . It would contain these properties:
+
+- Time that the query was run (required)
```

Comment:
This is somewhat of a meta point, but most of this discussion seems to be around an assumption that we are recording every _execution_ of a query as a unique instance in the database - so this idea of the time that the query was run makes sense in that context.  However, I could see wanting to use this to instead record every _unique_ query, with some information (like how often it is run) pre-computed before storing it in Amundsen.  For those cases, I don't know if this requirement really makes sense.